actionpack/lib/action_dispatch/middleware/flash.rb

```diff
@@ -82,15 +82,48 @@ def reset_session # :nodoc:
 
     module NullFlash #:nodoc:
       class << self
-        def []=(k, v); end
+        include Enumerable
 
         def [](k); end
+        def []=(k, v); end
 
+        def alert; end
         def alert=(message); end
 
+        def notice; end
         def notice=(message); end
 
-        def empty?; end
+        def empty?
+          true
+        end
+
+        def now
+          self
+        end
+
+        def update(h); end
+        alias :merge! :update
+
+        def replace(h); end
+
+        def keys; end
```

Comment:
I don't think that it's correct to have all these methods return `nil` like this. Sure it will now longer break if for instance you call `flash.keys`, but returning `nil` is totally unexpected and will likely break the caller.

NullObjects are notoriously hard to get right. I'm starting to wonder if returning a regular `Flash` object wouldn't be better. You'd just get a new one on every `request.flash` call. I doubt anyone has `request.flash`  as a hotspot so I don't think a few extra allocation are big concern. 

This will also be a good place to put the log suggested by @pixeltrix

Reply:
> [...] returning nil is totally unexpected and will likely break the caller.

Agree, I'll try to replicate the responses too. I also think that logs would be helpful.

> I'm starting to wonder if returning a regular Flash object wouldn't be better.

Well yes, the main issue in #42167 was that the Flash middleware wasn't present when some other modules (eg. `ETagWithFlash`) depended on it. The null object was added for the sake of correctness I would say, as even though it is present, it basically no-ops. A regular flash object works too, with the difference that it indeed would behave as if the flash module were working just fine.

Let me know if this null object still makes sense, so that I work on both items 😄.

Reply:
> Let me know if this null object still makes sense

Well it makes sense, but maintenance wise I think it's much more work than just returning an empty Flash. So I'd go with the later.

Reply:
Sounds good. I've removed the null object, now the request should have access to a regular (empty) flash object, even when the session store is disabled.

Btw, I was checking your branch the other day and tried to add the logging you mentioned: https://github.com/ricardotk002/rails/commit/587d599c4c

Something I noticed was that the Rails session always wrap the previous session into a regular enabled one, even if that previous one comes disabled (for instance, this is what I did for the tests). I'm not sure that'd be a common case but perhaps something to keep in mind: 

https://github.com/ricardotk002/rails/commit/587d599c4c#diff-40af0c0e76bdb69f4240deed5d844e4564115d291f8f73b0c4ebf523a1ed16bdR19-R23